---
title: "assignment9"
author: "Daniel Lowe"
date: "November 8th, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(stringi)
library(Hmisc)
library(UsingR)
library(tidyverse)
library(devtools)
library(rafalib)
library(qvalue)
library(genefilter)

library(dagdata)
data(admissions)
library(Biobase)
library(GSE5859)
data(GSE5859)
library(GSE5859Subset)
data(GSE5859Subset)
library(sva)
library(RColorBrewer)
```

# Basic Machine Learning

## Exercise 1

### 1. Let's compute the proportion of men who were accepted:

```{r}
index = which(admissions$Gender==0)
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100)
applied = sum(admissions$Number[index])
accepted/applied
```

What is the proportion of women that were accepted?

0.3033351 is the proportion of women that were accepted

### 2. Now that we have observed different acceptance rates between genders, test for the significance of this result.

```{r}
chi<-as.data.frame(matrix(2,2))
chi[1,1] <- sum(admissions[1:6,2]*admissions[1:6,3]/100)
chi[1,2] <- sum(admissions[1:6,2]*(1-admissions[1:6,3]/100))
chi[2,1] <- sum(admissions[7:12,2]*admissions[7:12,3]/100)
chi[2,2] <- sum(admissions[7:12,2]*(1-admissions[7:12,3]/100))
chisq.test(chi)$p.value
```

If you perform an independence test, what is the p-value?

An independence test gives a p value of  9.139492e-22

### 3. We can quantify how "hard" a major is by using the percent of students that were accepted.
Compute the percent that were accepted (regardless of gender) to each major and call this
vector H.

Which is the hardest major?

```{r}
asum <- (admissions[1,2]+admissions[7,2])
bsum <- (admissions[2,2]+admissions[8,2])
csum <- (admissions[3,2]+admissions[9,2])
dsum <- (admissions[4,2]+admissions[10,2])
esum <- (admissions[5,2]+admissions[11,2])
fsum <- (admissions[6,2]+admissions[12,2])

a <- admissions[1,2]/asum*admissions[1,3]/100 + admissions[7,2]/asum*admissions[7,3]/100
b <- admissions[2,2]/asum*admissions[2,3]/100 + admissions[8,2]/asum*admissions[8,3]/100
c <- admissions[3,2]/asum*admissions[3,3]/100 + admissions[9,2]/asum*admissions[9,3]/100
d <- admissions[4,2]/asum*admissions[4,3]/100 + admissions[10,2]/asum*admissions[10,3]/100
e <- admissions[5,2]/asum*admissions[5,3]/100 + admissions[11,2]/asum*admissions[11,3]/100
f <- admissions[6,2]/asum*admissions[6,3]/100 + admissions[12,2]/asum*admissions[12,3]/100
H <- c(a,b,c,d,e,f)
H
```

The hardest major is major is major F

### 4. What proportion is accepted for this major?

0.04957128 is the proportion of men accepted in this major

### 5. For men, what is the correlation between the number of applications across majors and H?

```{r}
cor(H,admissions[1:6,2])
```

The correlation is 0.8408551

### 6. For women, what is the correlation between the number of applications across majors and H?

```{r}
cor(H,admissions[7:12,2])
```

The correlation is -0.489443

### 7. Given the answers to the above, which best explains the differences in admission percentages when we combine majors?

. A) We made a coding mistake when computing the overall admissions percentages.
. B) There were more total number of women applications which made the denominator
much bigger.
. C) There is confounding between gender and preference for "hard" majors: females
are more likely to apply to harder majors.
. D) The sample size for the individual majors was not large enough to draw the correct
conclusion.

C best explains the difference in admission percentages

## Exercise 2

### 1. Familiarize yourself with the sampleInfo table. Note that some samples were processed at different times. This is an extraneous variable and should not affect the values in geneExpression. However, as we have seen in previous analyses, it does appear to have an effect so we will explore this here.

You can extract the year from each date like this:

```{r}
geneExpression = exprs(e)
sampleInfo = pData(e)

year = format(sampleInfo$date,"%y")
table(year,sampleInfo$ethnicity)
```

Note that ethnic group and year is almost perfectly confounded:

### 2. For how many of these years do we have more than one ethnicity represented?

2005 and 2006 have more than one ethnicity represented

### 3. Repeat the above exercise, but now, instead of year, consider the month as well. Specifically, instead of the year variable defined above use:

```{r}
month.year = format(sampleInfo$date,"%m%y")
table(month.year,sampleInfo$ethnicity)
```

For what proportion of these month.year values do we have more than one ethnicity
represented?

1 out of the 21 month.year values has more than 1 ethnicity represented

### 4. Perform a t-test (use rowttests) comparing CEU samples processed in 2002 to those processed in 2003. Then use the qvalue package to obtain q-values for each gene. How many genes have q-values < 0.05 ?

```{r}
year2 = factor(year)
tests = which(year2%in% c("02","03") & sampleInfo$ethnicity=="CEU")
year2 = droplevels(year2[tests])
p = rowttests(geneExpression[ ,tests], year2)$p.value
q = qvalue(p)
sum(q$qvalue < 0.05)
```

There are 4308 genes with q-values less than 0.05

### 5. What is the estimate of pi0 provided by qvalue:

```{r}
q$pi0
```

The pi0 estimate is 0.3628642

### 6. Now perform a t-test (use rowttests) comparing CEU samples processed in 2003 to those processed in 2004. Then use the qvalue package to obtain q-values for each gene. How many genes have q-values less than 0.05?

```{r}
year = format(sampleInfo$date,"%y")
year2 = factor(year)
tests = which(year2%in% c("03","04") & sampleInfo$ethnicity=="CEU")
year2 = droplevels(year2[tests])
p = rowttests(geneExpression[ ,tests], year2)$p.value
q = qvalue(p)
sum(q$qvalue < 0.05)
```

There are 2463 with q-values less than 0.05

### 7. Now we are going to compare ethnicities as was done in the original publication in which these data were first presented. Use the qvalue function to compare the ASN population to the CEU population. Once again, use the qvalue function to obtain q-values. How many genes have q-values < 0.05 ?

```{r}
race = factor( sampleInfo$ethnicity)
tests = which(race%in% c("CEU","ASN"))
race = droplevels(race[tests])
p = rowttests(geneExpression[ ,tests], race)$p.value
q = qvalue(p)
sum(q$qvalue < 0.05)
```

There are 7217 genes with q-values less than 0.05

### 8. Over 80% of genes are called differentially expressed between ethnic groups. However, due to the confounding with processing date, we need to confirm these differences are actually due to ethnicity. This will not be easy due to the almost perfect confounding. However, above we noted that two groups were represented in 2005. Just like we stratified by majors to remove the "major effect" in our admissions example, here we can stratify by year and perform a t-test comparing ASN and CEU, but only for samples processed in 2005. How many genes have q-values < 0.05 ?

```{r}
race = factor( sampleInfo$ethnicity)
year = format(sampleInfo$date,"%y")
year2 = factor(year)
tests = which(race%in% c("CEU","ASN") & year == "05")
race = droplevels(race[tests])
p = rowttests(geneExpression[ ,tests], race)$p.value
q = qvalue(p)
sum(q$qvalue < 0.05)
```

There are 560 genes with q-values less than 0.05

Notice the dramatic drop in the number of genes with q-value < 0.05 when we fix the year.
However, the sample size is much smaller in this latest analysis which means we have less
power:

```{r}
table(sampleInfo$ethnicity[index])
```



### 9. To provide a more balanced comparison, we repeat the analysis, but now taking 3 random CEU samples from 2002. Repeat the analysis above, but comparing the ASN from 2005 to three random CEU samples from 2002. Set the seed at 3, set.seed(3) How many genes have q-values < 0.05 ?

```{r}
set.seed(3)
race = factor( sampleInfo$ethnicity)
year = format(sampleInfo$date,"%y")
year2 = factor(year)
test1 = which(race%in% c("ASN") & year == "05")
test2 = which(race%in% c("CEU") & year == "02")
test2<-sample(test2,3)
tests <- c(test1,test2)
race = droplevels(race[tests])
p = rowttests(geneExpression[ ,tests], race)$p.value
q = qvalue(p)
sum(q$qvalue < 0.05)
```

There are 3695 with q-values less than 0.05

## Exercise 3

```{r}
data(GSE5859Subset)
sex = sampleInfo$group
month = factor( format(sampleInfo$date,"%m"))
table(  sampleInfo$group, month)
```

### 1. Using the functions rowttests and qvalue compare the two groups. Because this is a smaller dataset which decreases our power, we will use the more lenient FDR cut-off of 10%. How many gene have q-values less than 0.1?

```{r}
gender = factor(sex)
p = rowttests(geneExpression,gender)$p.value
q = qvalue(p)
sum(q$qvalue < 0.1)
```

There are 59 genes with a q-values less than 0.1

### 2. Note that sampleInfo$group here presents males and females. Thus, we expect differences to be in on chrY and, for genes that escape inactivation, chrX. We do not expect many autosomal genes to be different between males and females. This gives us an opportunity to evaluate false and true positives with experimental data. For example, we evaluate results using the proportion genes of the list that are on chrX or chrY. For the list calculated above, what proportion of this list is on chrX or chrY?

```{r}
gender = factor(sex)
p = rowttests(geneExpression,gender)$p.value
q = qvalue(p)$qvalue
genes = geneAnnotation$CHR[q<0.1]
genes2 = genes[genes == "chrX" | genes == "chrY"]
length(genes2)/length(genes)
```

0.3389831 is the proportion of this list on chrY or chrY

### 3. We can also check how many of the chromosomes X and Y genes we detected as different. How many are on Y?

```{r}
genes3 = genes[genes == "chrY"]
length(genes3)
```

8 of the genes are on Y

### 4. Now for the autosomal genes (not on chrX and chrY) for which q-value < 0.1, perform a t-test comparing samples processed in June to those processed in October. What proportion of these have p-values <0.05 ?

```{r}
gender = factor(sex)
p = rowttests(geneExpression,gender)$p.value
q = qvalue(p)$qvalue
genes = which(q<0.1 & !geneAnnotation$CHR%in%c("chrX","chrY"))
m = factor( format(sampleInfo$date,"%m"))
p = rowttests(geneExpression[genes,],month)$p.value
length(p[p<0.05])/length(p)
```

0.8717949 of these have a p-value less than 0.05

### 5. The above result shows that the great majority of the autosomal genes show differences due to processing data. This provides further evidence that confounding is resulting in false positives. So we are going to try to model the month effect to better estimate the sex effect. We are going to use a linear model:

Which of the following creates the appropriate design matrix?
. A) X = model.matrix(???sex+ethnicity)
. B) X = cbind(sex,as.numeric(month))
. C) It can't be done with one line.
. D) X = model.matrix(???sex+month)

The correct answer is D, X = model.matrix(???sex+month)

### 6. Now use the X defined above, to fit a regression model using lm for each gene. You can obtain p-values for estimated parameters using summary. Here is an example

```{r}
X = model.matrix(~sex+month)
i = 234
y = geneExpression[i,]
fit = lm(y~X)
summary(fit)$coef
head(geneExpression)
```

How many of the q-values for the group comparison are now <0.1?

```{r}
result <- vector("integer", 8793)
for (i in 1:8793){
  X = model.matrix(~sex+month)
  y = geneExpression[i,]
  fit = lm(y~X)
  result[i] <- summary(fit)$coef[2,4]
}
```

```{r}
q = qvalue(result)
sum(q$qvalues<0.1)
```

17 qvalues for the group comparison are less than 0.1

### 7. With this new list, what proportion of these are chrX and chrY?

```{r}
genes <- geneAnnotation$CHR[q$qvalues<0.1]
genes2 <- genes[genes == "chrX" | genes == "chrY"]
length(genes2)/length(genes)
```

0.8823529 is the proportion of these are chrX and chrY

### 8. How many on Y or X?

```{r}
length(genes[genes == "chrY"])
length(genes[genes == "chrX"])
```

6 are on Y, 9 are on X

### 9. Now from the linear model above, extract the p-values related to the coefficient representing the October versus June differences using the same linear model. How many of the q-values for the month comparison are now <0.1? This approach is basically the approach implemented by Combat.

```{r}
result <- vector("integer", 8793)
for (i in 1:8793){
  X = model.matrix(~sex+month)
  y = geneExpression[i,]
  fit = lm(y~X)
  result[i] <- summary(fit)$coef[3,4]
}

q = qvalue(result)
sum(q$qvalues<0.1)
```

3170 genes have a qvalue less than 0.1 when the month comparison is used

## Exercise 4

### 1. Suppose you want to make an MA plot of the first two samples y = geneExpression[,1:2]. Which of the following projections gives us the projection of y so that column2 versus column 1 is an MA plot?

A,B,C,D?

The correct answer is B, y(1,1;1,-1)

### 2. Say Y is M × N, in the SVD Y = UDV ??? which of the following is not correct?

. A) DV ??? are the new coordinates for the projection U???Y
. B) UD are the new coordinates for the projection Y V
. C) D are the coordinates of the projection U???Y
. D) U???Y is a projection from an N-dimensional to M-dimensional subspace.

C, D are the coordinates of the projection U???Y, is not correct

### 3. Define: Compute and plot an image of the correlation for each sample. Make two image plots of these correlations. In the first one, plot the correlation as image. In the second, order the samples by date and then plot an image of the correlation. The only difference in these plots is the order in which the samples are plotted.

```{r}
y = geneExpression - rowMeans(geneExpression)
n=ncol(y)
cors=cor(y)
cols=colorRampPalette(rev(brewer.pal(11,"RdBu")))(100)
mypar()
image(1:n,1:n,cors,xaxt="n",yaxt="n",col=cols,xlab="",ylab="",zlim=c(-1,1))
```

```{r}
order = order(sampleInfo$date)
y1 = y[,order]
cors=cor(y1)
cols=colorRampPalette(rev(brewer.pal(11,"RdBu")))(100)
mypar()
image(1:n,1:n,cors,xaxt="n",yaxt="n",col=cols,xlab="",ylab="",zlim=c(-1,1))
```

Based on these plots, which of the following you would say is true?

. A) The samples appear to be completely independent of each other.
. B) Sex seems to be creating structures as evidenced by the two cluster of highly
correlated samples.
. C) There appear to be only two factors completely driven by month.
. D) The fact that in the plot ordered by month we see two groups mainly driven by
month, and within these we see subgroups driven by date, seems to suggest date more
than month per se are the hidden factors.

The correct answer is D, The fact that in the plot ordered by month we see two groups mainly driven by
month, and within these we see subgroups driven by date, seems to suggest date more
than month per se are the hidden factors.

### 4. Based on the correlation plots above, we could argue that there are at least two hidden factors. Using PCA estimate these two factors. Specifically, apply the svd to y and use the first two PCs as estimates. Which command gives us these estimates?

. A) pcs = svd(y)$v[1:2,]
. B) pcs = svd(y)$v[,1:2]
. C) pcs = svd(y)$u[,1:2]
. D) pcs = svd(y)$d[1:2]

The correct answer is B, pcs = svd(y)$v[,1:2]

### 5. Plot each of the estimated factors ordered by date. Use color to denote month. The first factor is clearly related to date. Which of the following appear to be most different according to this factor?

```{r}
month=format(sampleInfo$date, "%m")
pcs = svd(y)$v[,1:2]
o = order(sampleInfo$date)
cols = as.numeric(month)[o]
plot(pcs[o,1],col=cols,xaxt="n",xlab="")
label = gsub("2005-","",sampleInfo$date[o])
axis(1,1:ncol(y),label,las=2)
```

. A) June 23 and June 27
. B) Oct 07 and Oct 28
. C) June 10 and June 23
. D) June 15 and June 24

The correct answer is A, June 23 and June 27

### 6. Use the svd function to obtain the principal components (PCs) for our detrended gene expression data y. How many PCs explain more than 10% of the variability?

```{r}
s = svd(y)
PC = s$d^2/ sum(s$d^2)
sum(PC>0.10)
```

2 PCs explain more than 10% of the variability

### 7. Which PC most correlates (negative or positive correlation) with month?

```{r}
month = factor( format(sampleInfo$date,"%m"))
cors = cor( as.numeric(month),s$v)
cors
max(abs(cors))
```
 
 The first component most correlates with month
 
### 8. What is this correlation (in absolute value)?

This correlation is 0.8297915

### 9. Which PC most correlates (negative or positive correlation) with sex?

```{r}
sex= sampleInfo$group
cors = cor( as.numeric(sex),s$v)
cors
max(abs(cors))
```

The first principle component most correlates with sex

### 10. What is this correlation (in absolute value)?

This correlation is 0.6236858

### 11. Now instead of using month, which we have shown does not quite describe the batch, add the two estimated factors s$v[,1:2] to the linear model we used above. Apply this model to each gene and compute q-values for the sex difference. How many q-values < 0.1 for the sex comparison?

```{r}
result <- vector("integer", 8793)
for (i in 1:8793){
  X <- model.matrix(~sex+s$v[,1:2])
  y = geneExpression[i,]
  fit = lm(y~X)
  result[i] <- summary(fit)$coef[2,4]
}
q = qvalue(result)
sum(q$qvalues<0.1)
```

There were 14 qvalues less than 0.1 for the sex comparision

### 12. What proportion of the genes are on chromosomes X and Y?

```{r}
genes = geneAnnotation$CHR[qvals<0.1]
genes2 = genes[genes == "chrX " | genes == "chrY"]
length(genes2)/length(genes)
```

0.0148265 is the proportion of the genes on chromosomes X and Y

## Exercise 5

### 1. In a previous section we estimated factors using PCA, but we noted that the first factor was correlated with our outcome of interest:

The svafit function estimates factors, but downweighs the genes that appear to correlate
with the outcome of interest. It also tries to estimate the number of factors and returns the
estimated factors like this:

The resulting estimated factors are not that different from the PCs.

```{r}
s <- svd(geneExpression-rowMeans(geneExpression))
cor(sampleInfo$group,s$v[,1])

sex = sampleInfo$group
mod = model.matrix(~sex)
svafit = sva(geneExpression,mod)
head(svafit$sv)

for(i in 1:ncol(svafit$sv)){
print( cor(s$v[,i],svafit$sv[,i]) )
}
```

Now fit a linear model to each gene that instead of month includes these factors in the model.
Use the qvalue function.
How many genes have q-value < 0.1?



### 2. How many of these genes are from chrY or chrX?


